fix: NetcodeIntegrationTest destroy GameObjects not NetworkObjects during shutdown sequence

NoelStephensUnity opened the pull request:
This fixes a minor oversight in the NetcodeIntegrationTest where the DestroySceneNetworkObjects method was destroying the NetworkObject component and not the GameObject the NetworkObject component is assigned to.

[MTT-2686](https://jira.unity3d.com/browse/MTT-2686)

## Testing and Documentation
* No tests have been added.

NoelStephensUnity committed:
fix

Destroy the GameObject and not the NetworkObject during DestroySceneNetworkObjects

NoelStephensUnity committed:
style

Added comment.

NoelStephensUnity committed:
fix

Some scenarios the NetworkObject and/or the GameObject can be null.
This update takes that into account.

mattwalsh-unity commented on com.unity.netcode.gameobjects/TestHelpers/Runtime/NetcodeIntegrationTest.cs:
```diff
@@ -493,9 +493,17 @@ protected void DestroySceneNetworkObjects()
+                    if (networkObject.gameObject != null)
+                    {
+                        // Destroy the GameObject that holds the NetworkObject component
+                        Object.DestroyImmediate(networkObject.gameObject);
```
Why would you not want to also do `Object.DestroyImmediate(networkObject);` here?

NoelStephensUnity replied:
Because when you destroy a GameObject all attached components and children of the GameObject are destroyed along with it.  The issue being it was just destroying the NetworkObject which left the GameObject (and remaining components) still instantiated.

mattwalsh-unity replied:
Ah, I see.  Well, let's add comments of "I have an orphaned NO, delete it" and "in deleting the GO all the behaviours are cleaned up also)."

Though I am a bit confused of...
- seems very odd to me that you could have an orphaned NetworkObject.
- ...though is this caused by having `CanDestroyNetworkObject` return true?  Do we like that?  Is it better to just have NO's that don't persist?

NoelStephensUnity replied:
Adjusted the code to better reflect this scenario and added comments to explain how a NetworkObject in a list of found NetworkObjects can be null.

NoelStephensUnity committed:
refactor fix

This is a cleaner and more accurate fix for the issue where a parent is destroyed before its child/children.  Added additional comments to explain why.
